tensorflow/compiler/tf2xla/xla_compiler.cc

```diff
@@ -234,6 +234,9 @@ Status BuildComputation(
       }
 
       case XlaExpression::Kind::kResource:
+        // resources are pushed into elems later when processing resource
+        // arguments. This is correct as long as the input and output resources
```

Comment:
How do we know that this is true:

```
        // This is correct as long as the input and output resources
        // are in the same order.
```
?

Reply:
Because:
1. outputs for resources are always created when return_updated_values_for_all_resources is set to true.
2. the resource output retvals are sorted according to the input argument position ([here](https://github.com/tensorflow/tensorflow/blob/408cea8c638a9873c8a23887fadf74314bd2a0b9/tensorflow/compiler/tf2xla/xla_compiler.cc#L258))
3. So, as long as the outputs are in the same order as the inputs, they will be pushed into elems in the correct order ([here](https://github.com/tensorflow/tensorflow/blob/408cea8c638a9873c8a23887fadf74314bd2a0b9/tensorflow/compiler/tf2xla/xla_compiler.cc#L304)).

Reply:
> So, as long as the outputs are in the same order as the inputs

I meant what guarantees that the `DT_RESOURCE` outputs are in the same order as the inputs?

Reply:
Ah. Got it. It is guaranteed by the functionalization while pass, since it always adds a corresponding output to a DT_RESOURCE input.

Reply:
Ok, great, can you add this to the comment?

Reply:
BTW, I guess you may also ask in terms of cases other than tf.while (although this commit changes nothing about them). My understanding is as follows--functions other than the while loop body do not have DT_RESOURCE as outputs when they are generated by TF. So, this case may never be hit by them. I, however, wonder why the original author of the codes does not choose to bail out in this case. Do you have any insight?

Reply:
Added.

Reply:
> Ah. Got it. It is guaranteed by the functionalization while pass, since it always adds a corresponding output to a DT_RESOURCE input.

But it isn't guaranteed for `While` loops generated directly by TF.  Right now we have this `RearrangeFunctionArgumentPass` that (I think unsoundly, but I'm double checking with the author) rewrites TF `While` loop bodies to *not* return `DT_RESOURCE`s (so making it look like what the old functionalization pass would have generated).  But we'll have to disable that pass when we check this in since otherwise we won't do the right thing for TF generated while loops.

> My understanding is as follows--functions other than the while loop body do not have DT_RESOURCE as outputs when they are generated by TF. So, this case may never be hit by them.

I agree -- we've just been getting lucky because TF happens not to generate such functions.  But IMO we should at least return an `UnimplementedError` instead of silently miscompiling if such odd functions do show up.

Reply:
Thank your for bringing up the potential issue related to `RearrangeFunctionArgumentPass` for discussions. I read the code in the pass and summarize my current understanding as follows. Let's see if they are aligned to your understanding.

1. `xla_compiler` supports the following two forms of TF functions.
  a. `DT_RESOURCE` are only inputs and placed in the last.
  b. `DT_RESOURCE` are both inputs and outputs. They still need to be placed in the last and the input/output order needs to be aligned.
2. TF can generate functions that do not conform to the two forms listed in item 1, but `RearrangeFunctionArgumentPass` will fix them by rewriting the functions to form 1.a. However, note that it rewrites the while loop body **only if** DT_RESOURCE arguments are not placed in the last.
3. Based on 2, `RearrangeFunctionArgumentPass` will not rewrite the functionalized while, as its DT_RESOURCE arguments are always placed in the last.
4. We should not turn off `RearrangeFunctionArgumentPass`, as it is needed to fix functions generated from TF. Also, it does not affect the functionalized while.

Reply:
> Thank your for bringing up the potential issue related to `RearrangeFunctionArgumentPass` for discussions. I read the code in the pass and summarize my current understanding as follows. Let's see if they are aligned to your understanding.
> 
> 1. `xla_compiler` supports the following two forms of TF functions.
>    a. `DT_RESOURCE` are only inputs and placed in the last.
>    b. `DT_RESOURCE` are both inputs and outputs. They still need to be placed in the last and the input/output order needs to be aligned.
> 2. TF can generate functions that do not conform to the two forms listed in item 1, but `RearrangeFunctionArgumentPass` will fix them by rewriting the functions to form 1.a. However, note that it rewrites the while loop body **only if** DT_RESOURCE arguments are not placed in the last.

Moreover, 1a. for while loops is not correct.

The behavior of a while loop is, roughly, `while (cond(state)) { state = body(state); }` which means the input and output state from the while loop body needs to have the same shape.  So the while loop body cannot be edited to remove the `DT_RESOURCE` outputs -- if they're present in the inputs, the need to be present in the outputs.  In that sense `RearrangeFunctionArgumentPass` is unsound because it will convert a correct While loop generated by TF into an incorrect while loop.  Now these incorrect while loops with "work" when run through the XLA compiler because the XLA compiler always places `DT_RESOURCE` outputs at the end of the output tuple from a function call, even if the function did not have those `DT_RESOURCE` outputs.

> 3. Based on 2, `RearrangeFunctionArgumentPass` will _not_ rewrite the functionalized while, as its DT_RESOURCE arguments are always placed in the last.

I agree.

> 4. We should not turn off `RearrangeFunctionArgumentPass`, as it is needed to fix functions generated from TF. Also, it does not affect the functionalized while.

We will still need the "rearrange" aspect of `RearrangeFunctionArgumentPass`.  However, once your change is checked in, we can change the XLA compiler to be stricter about the While loops it accepts (possibly as an explicit `TF_RET_CHECK` or something) by ensuring that the output shape of the TF body computation matches its input shape (*you* don't have to do this, but that would be a logical change to make).  Then the XLA compiler would start failing with while nodes generated by TF and mangled by the `RearrangeFunctionArgumentPass` pass.  So `RearrangeFunctionArgumentPass` needs to be fixed in the same change that makes the XLA compiler stricter.

> But we'll have to disable that pass when we check this in

I was wrong here.  I don't think we need to disable `RearrangeFunctionArgumentPass` when we check *this* in, but we'll need to disable it when we make XLA compiler stricter as I suggested above.

Reply:
Just to close this off, I think the current code is fine, but it would be nice to have future CLs that:

 - Add a `TF_RET_CHECK(retval.resource()->arg_num() == i)` or equivalent here.
 - Checks that resources, if returned from the TF function, are in the correct order.
 - Makes the `XlaOpKernel` for `While` stricter around mismatching input/output shapes for the body computation.  `RearrangeFunctionArgumentPass` will have to be fixed in this CL.

The first two invariants should always hold because of `RearrangeFunctionArgumentPass` and how we functionalize while loops, but having an explicit check will make the code easier to understand and debug.

If you want to take these on then that's great, otherwise please let me know and I or someone else from my team can figure these out.

Reply:
No problem. I will be happy to submit a CL fixing them.